lib/ops/opsservice/authgateway.go

```diff
@@ -45,7 +48,12 @@ func (o *Operator) UpsertAuthGateway(key ops.SiteKey, gw storage.AuthGateway) er
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	return UpsertAuthGateway(client, o.users(), gw)
+	err = UpsertAuthGateway(client, o.users(), gw)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	events.Emit(ctx, o, events.AuthGatewayUpdated, events.Fields{})
```

Comment:
nit: Maybe you should have added another API - `EmitWithFields` so that the empty `Fields` would not be necessary?
Another thought is why not keep event and its fields together? This way it would be natural both to construct new events:
```go
// this one has the benefit of allowing an empty constructor:
events.Emit(ctx, o, events.NewTLSKeyPairCreated())
events.Emit(ctx, o, events.NewUserCreated(events.FieldName, user.GetName())
// or
events.Emit(ctx, o, events.NewUserCreated(events.Fields{events.FieldName: user.GetName()})
```
then:
[lib/ops/events/events.go]
```go
type Event struct {
	events.Event
	events.Fields
}
```
and emit them.

Reply:
I'd really like to avoid having "constructors" for each event type. That was the approach I took originally and it didn't bring any benefits, only 500+ lines of paperweight code b/c we have so many events. It should be fine to emit just the event struct + fields. I'll see about having "Emit" and "EmitWithFields" though, that sounds reasonable.